src/cat.jl

```diff
@@ -38,14 +38,14 @@ function hcat(A::AbstractLinearOperator, B::AbstractLinearOperator)
   nrow = size(A, 1)
   Ancol, Bncol = size(A, 2), size(B, 2)
   ncol = Ancol + Bncol
-  S = promote_type(eltype(A), eltype(B))
+  T = promote_type(eltype(A), eltype(B))
   prod! = @closure (res, v, α, β) -> hcat_prod!(res, A, B, Ancol, Ancol + Bncol, v, α, β)
   tprod! = @closure (res, u, α, β) ->
     hcat_ctprod!(res, transpose(A), transpose(B), Ancol, Ancol + Bncol, u, α, β)
   ctprod! = @closure (res, w, α, β) ->
     hcat_ctprod!(res, adjoint(A), adjoint(B), Ancol, Ancol + Bncol, w, α, β)
   args5 = (has_args5(A) && has_args5(B))
-  CompositeLinearOperator(S, nrow, ncol, false, false, prod!, tprod!, ctprod!, args5)
+  CompositeLinearOperator(T, nrow, ncol, false, false, prod!, tprod!, ctprod!, args5, S = storage_type(A))
```

Comment:
This is odd. Suppose `A` uses vectors on the CPU and `B` on the GPU. Now if `C = [A B]`, what should be the storage type of C? Why should it be the same as `A`? How is a product `C * x` supposed to work? Should the `A` part occur on the CPU and the `B` part on the GPU?

Reply:
The goal of this PR is to be able to work with Arrays on the GPU only. It looks like `mul!(res, A, x)` does not work if `res` is on the CPU and `A`, `x` a matrix and a vector on the GPU, so this would raise issues. If possible, I suggest this behaviour should also be considered in another PR.

Reply:
In that case, the code should raise an error (in this PR) if the two operators have different "storage types".